vllm-project/vllm
# Support int8 KVCacheQuant and W8A8 inference in vllm

We have recently implemented and tested int8 KV-Cache quantization and W8A8 inference in vLLM. We find out that our quantization implementation can **increase the throughput over 20%** and **reduce the first token latency** under heavy load. In contrast,  W4A16 quant methods(eg. AWQ-based method) provided in vllm cannot improve throughput according to [pr1032](https://github.com/vllm-project/vllm/pull/1032) because it can not benefit from int8 tensor core.  So we propose this PR as an alternative quantization method.

***Updates!!!***
We have made some more progress in https://github.com/vllm-project/vllm/pull/1112#issuecomment-1770889034

***More Updates!!!***
If you want to  properly eval mmlu dataset with vllm, some modify on sampler must be done. The code can be found in our [mmlu_eval](https://github.com/AniZpZ/vllm/tree/mmlu_eval) branch

***Important message!!!***
We spilt the pr into two parts for easier review and use. The w8a8 inference part is in https://github.com/vllm-project/vllm/pull/1508 and the kv cache quant part is in https://github.com/vllm-project/vllm/pull/1507


**What we have right now:**
1.  int8 KV-Cache quantization related works:
   a. Quant\Dequant helper functions adapted from Faster Transformer
   b. Quantized version CUDA kernels 
   c. Unit tests for the added kernels
2. W8A8 inference related works:
   a. Int8 Gemm kernels adapted from torch-int
   b. W8A8 linear layer modules
   c. Support W8A8 inference on Llama model
3. Test result based on our own dataset

**What we plan to do:**

- [x] 1. Further kernel fusion
- [x] 2. Code refactoring and cleaning
- [x] 3. Opimize int8 GEMM kernel
- [x] 4. Release SmoothQuant for LLaMA
- [x] 5. Add code for generating KV-Cache quantization parameters (scales and zero points)
- [x] 6. Experiments on more datasets



**How to test throughput**
A. how to enable w8a8 inference
~~0. install cutlass because we currently use  cutlass gemm kernel. We plan to replace them with cublas gemm kernel soon.~~
We support cublas gemm kernel now, you can remove cutlass gemm kernel in setup.py
1. install [smoothquant](https://github.com/AniZpZ/smoothquant/tree/llama-dev) and [torch-int](https://github.com/AniZpZ/torch-int/tree/llama-dev) for llama. Use "examples/generate_act_scales.py" to generate act scale, and then use "examples/export_int8_llama.py" to export int8 model. Please note to check and change the 'architectures' field in the model's config.json from 'Int8LlamaForCausalLM' to 'LlamaForCausalLM'.
2. update vllm and execute
```
python ./vllm/entrypoints/api_server.py --model=/path/to/quantized/model --tokenizer=/path/to/tokenizer --max-num-batched-tokens=70000 --block-size=16 --swap-space=20 --quantization smoothquant
``` 

B. how to enable kv cache quant
1. use vllm/kv_quant/calibrate.py to genearte scales and use vllm/kv_quant/export_kv_params.py to export kv caches.
2. exeute
```
python ./vllm/entrypoints/api_server.py --model=/path/to/quantized/model --tokenizer=/path/to/tokenizer --max-num-batched-tokens=70000 --block-size=16 --swap-space=20 --kv-cache-dtype=int8 --kv-quant-params-path=/path/to/kv_params_dir
```  
And you can use kv cache quant and w8a8 inference together



**Experiment Result** 
current test result in our datasets on A100 80G (updates with quant&rms fusion and gemm d2h bug fix)

Throughput of FP16 LLaMA-13B:
```
Throughput:  4.9945 requests/s, 543.0660 token/s
Average latency: 31.7689 s
``` 

Throughput of Int8 LLaMA-13B with int8 KVCacheQuant:
```
Throughput: 6.1147 requests/s, 664.8646 token/s, 
Average latency: 27.4222 s
``` 
Throughput of Int8 LLaMA-13B with int8 KVCacheQuant, using cublas gemm kernel:
```
Throughput: 6.4723 requests/s, 703.7514 token/s, 
Average latency: 25.9912 s
``` 



**How to evalute model performance**
We add evaluation method of quanted models, currently support mmlu datasets.
You can find detail in benchmarks/benchmark_evaluation.py
```
python benchmark_evaluation.py --model=/path/to/quantized/model --tokenizer=/path/to/tokenizer --dev-data-path=/path/to/mmlu/dev/ --test-data-path=/path/to/mmlu/test/ --kv-cache-dtype=int8 --kv-quant-params-path=/path/to/kv_params_dir --quantization=smoothquant
``` 



**Updates**
We have released SmoothQuant for LLaMA in 
https://github.com/AniZpZ/smoothquant/tree/llama-dev 
https://github.com/AniZpZ/torch-int/tree/llama-dev

The code for generating KV-Cache quantization parameters is ready, check vllm/kv_quant fold

We replace int8 gemm with cublas version and the **increasement of throughput comes to around 30%**
